### DISCUSSION TITLE
PLDI 2023 - Collected notes and conversations

### DISCUSSION THREAD

**Discussion by saulshanabrook** [saulshanabrook/saulshanabrook#27] [Blog]
Phew, what a week! I just finished attending the ["44th ACM SIGPLAN Conference on Programming Language Design and Implementation (PLDI 2023)"](https://pldi23.sigplan.org/). 
  
I had a great time, met lots of people, and heard about new interesting ideas. I wanted to write up some notes, to try to remember what came out of it, and share it with others. It took me a little longer than I expected to get to this, so it's not as fresh in my memory as I would like. If I got anything wrong please LMK and I will correct it!

---

# Egglog Stuff
There were a number of [egglog](https://github.com/egraphs-good/egglog) conversations during the conference.

## My Talk

I gave  [a talk on an egraphs in Python](https://pldi23.sigplan.org/details/egraphs-2023-papers/5/Egg-smol-Python-A-Pythonic-Library-for-E-graphs) at the egraphs workshop. It seemed to be well received over all. Most of the questions where around the use case in Python around data science APIs. I tried to record it, but wasn't able to unfortunately.

## Python in the Browser

@tonyfast suggested it would be interesting to try compiling egglog Python to WASM so we could play with it interactively in thr browser.
  

## Add PyObject
Mentioned to @philzook58 that I was interested in adding support for `PyObject`s as a builtin sort to the Python bindings. This could be used for embedding things like Python AST objects, bytecode objects, or other runtime objects in the e-graph, and converting between them and normalized e-graph forms, all with replacement rules.

@kszucs [also mentioned](https://github.com/ibis-project/ibis/pull/5781#issuecomment-1486569924) looking for this feature when trying to evaluate the project for use in Ibis as the replacement engine.

I don't think it would be too difficult. I would one additional builtin sort, which would store a pointer to a pyobject. Then the only function I would really have to add would be some form of `exec`, taking some string, the globals and locals. [Here](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.run) is how to use it from Rust.  

We did something [similar in lineapy](https://github.com/LineaLabs/lineapy/blob/872c30474277665aed22d8566308c5c334f417ab/lineapy/utils/lineabuiltins.py#L142-L202), adding `exec` and `exec_expr` nodes. A couple of things to think about is if/how to support custom globals/locals as well as source location for tracebacks.

Then I would also want to add a couple of conversion functions to/from PyObjects to the other builtin sorts.

From the Python interface side, I  would want to wrap this so users could define rewrite rules easily, without having to manually create strings. Something like this:

  

```python
"""
Compile from Python AST object to normalized expression to allow optimizations
"""

from __future__ import annotations
import ast
from typing import ClassVar
from egglog import *

# Parse Python AST into expression
egraph = EGraph()

@egraph.class_
class PyASTExpr:
    def __init__(self, py_ast: ast.Expression):
		...

@egraph.function
def constant(x: object) -> PyASTExpr:
	"""
	https://docs.python.org/3/library/ast.html#ast.Constant
	"""
	...

@egraph.class_
class BinOp:
	"""
	https://docs.python.org/3/library/ast.html#ast.Add
	"""	
	ADD: ClassVar[BinOp]

@egraph.function
def bin_op(left: PyASTExpr, op: BinOp, right: PyASTExpr) -> PyASTExpr:
	"""
	https://docs.python.org/3/library/ast.html#ast.BinOp
	"""
	...

x, y = vars_("x y", object)


# Add the ability to have Python code in a rewrite rule
egraph.register(
	replace(
		bin_op(constant(x), BinOp.ADD, constant(y))
	).with_(
		constant(eval_(lambda: x + y))
	)
)

# This would be syntactic sugar for something like this:
egraph.register(
	replace(
		bin_op(constant(x), BinOp.ADD, constant(y))
	).with_(
		constant(pyobject_eval("x + y", "x", x, "y", y))
	)
)
```

  
Would also have to add ways to go from PyObject bool to unit type, so could use result of Python object evals as conditionals for rules.
  
Have to also nail down the typing differences on the annotation side. At runtime, all egglog PyObject instance will be the same sort. But for the Python annotations, might want to add more granular annotations (like `ast.Expression`). Need to be careful to differentiate `int` as a `PyObject` and `int` converted to an `i64`. We could do it explicitly using [`Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) if we want like:

  

```python
T = TypeVar("T")
PyObject = Annotated[T, "PyObject"]

@egraph.class_
class PyASTExpr:
	def __init__(self, py_ast: PyObject[ast.Expression]):
		...
```

  
So that when inspecting the annotations we can see which are supposed to PyObjects. But this is a bit more cumbersome for users.

## Graphviz PR
  
Chatted with @oflatt about the graphviz PR (https://github.com/egraphs-good/egglog/pull/147). The main questions he had was on [the exported IR data structure](https://github.com/saulshanabrook/egg-smol/blob/daa97afbb93158225690f28e4670c574902fb68d/src/graph/mod.rs), since he was working on a similar one in an unrelated PR. I mentioned that I liked compiling to this IR in between the e-graph and the graphviz, mainly for mental hygiene and understanding. Also that we could eventually maybe expose this, or something like it, so that from the Python bindings I could combine to a different graph viz frontend, like cytoscapejs.
  

## Difference between primitives and e-classes

Chated with @ezrosent a bit about how that the fact that primitive sorts and user defined (e-class) sorts in egglog behave differently. I.e. you can call union on functions which return e-classes but not on those that return primitives. Also if a function with returns a primitive does not return a default and doesn't have a value set for some args, you cannot call it. This is not true for functions which return e-classes, since they will always return the e-class id. Which makes sense due to the implementation, but conceptually it's a bit odd, since some functions feel "eager" and others "lazy."

I was saying that I thought it would be great to rectify it, and make them all behave a bit more like e-classes, but I was unsure about the merge attribute, since it was essentially destructive, how that would work.

Eli said that he was thinking it could work in a way so that if a function return a primitive and defines a merge function, then it would be resolved in such a way that the result could have arbitrary number of e-nodes but only one primitive. And that the destructive nature could be implemented using delete in rule actions.  

I was super happy to hear this, since I had also been independently thinking a design like that would make the most sense. If this was changed, I think it would make egglog a lot more straightforward to use.

  

## Lambda Calculus

There was an interesting talk on how to support an embedded lambda calculus in e-graphs, which focused on different techniques to avoid the explosion of nodes which can result when you push down the substitution function recursively through pattern matching.

I talked with @Bastacyclop about his approach in his paper ["Sketch-Guided Equality Saturation"](https://arxiv.org/abs/2111.13040) (Section 4 "EFFICIENT EQUALITY SATURATION FOR THE LAMBDA CALCULUS") , how he tried different ways of doing lambda evaluation, outside of the builtin pattern matching framework. That he got to some thorny cases due to the possibly cyclical nature of the egraphs and handling higher order functions.

  

There is a [current lambda calculus](https://github.com/egraphs-good/egglog/blob/187441ad0d6826e8228b833dd41d384bc571c847/tests/lambda.egg) example which relies entirely on syntactic rewrites in egglog. One issue is definitely that it blows up the egraph size. But more significantly, it requires explicit matching on any subterm function, so cannot be composed arbitrary with other functions. This is because there is no way to match on a generic function in egglog, only on particular ones.

  

In metadsl, I originally did something like that example, but allowed matching on any function. However, it ended up being too efficient to push down the pattern matching through replacement rules. So instead, I ended up refactoring by [doing the replacement rewrite all in one rule](https://github.com/metadsl/metadsl/blob/56e21588602bc47c6eea4d81c7cbab41834a4f2c/metadsl_core/abstraction.py#L32-L137).

  

I wanted to try something similar in egglog, so I worked on a proof of concept PR (https://github.com/egraphs-good/egglog/pull/155) which would do the application in one go, by traversing the e-graph directly.

  

FWIW, I think eventually if egglog was to add something like this, it should be added with some more syntactic sugar, so that defining the vars are not as cumbersome. Also, at the type level, it would be great if builtin/user defined functions would be the same type as these lambda functions, so that functions like map could take either.

  

Reflecting on my metadsl implementation, this would be my main large regret about lambdas. There, like the current egglog PR, lambdas and application act just like any other user defined type, there is nothing special about them, and they exist on a different type level than functions. This adds some mental overhead and confusion. So instead, having them be unified at the type level would be ideal I think.

## LLVM

I talked with @ztatlock about how we could use e-graphs for compiling to LLVM. 

When chatting about what sort of IR to use to represent LLVM, he pointed me to the work he had done earlier on [Peggy](https://goto.ucsd.edu/~mstepp/peggy/). They have a paper on ["Equality-Based Translation Validator for LLVM"](https://goto.ucsd.edu/~mstepp/peggy/pubs/cav2011.pdf). 

![Screenshot 2023-07-03 at 4 01 55 PM](https://github.com/saulshanabrook/saulshanabrook/assets/1186124/7d36d02e-1921-4541-853a-01979d6b2e43)

I would be interested in talking more about how to design a system that goes from imperative (say LLVM or Python bytecode) to functional and then back to imperative. I am curious if that is at all possible within an e-graph framework. It seems like a very fertile area to experiment, because then you can start trying different encodings for the functional IR to see which are expressive enough and also have good optimization properties.

This is very related to the chat around mixed embeddings. 

## Type and Sorts

On the last point, I had a bunch of discussions with different folks about what sort of type system egglog should have. It isn't a super open design question, it seems pretty clear it won't be dependent types, but some form of System F AKA Hindley-Milner.  

In Python land, I will have users annotate every free variable (like in rewrites) with a type, and of course, all function signatures are explicitly annotated, which makes inference relatively trivial I believe?

Being able to defined use defined structures like cons lists, pairs, etc, would be great and depend on having some form of polymorphism.

For me, this comes up with things like implementing the mathematics of arrays and also for any type of monadic state IR for program compilation.

If we are thinking in terms of the Lambda cube, currently the type system is basically a [simply typed lambda calculus](https://en.wikipedia.org/wiki/Lambda_cube#(%CE%BB%E2%86%92)_Simply_typed_lambda_calculus).

If we want to add:

* terms which depend on types (for example, the `var` function in the lambda calculus PR I added needs to depend on a type, and return a term of that type)
* Types that depend on types (for example a cons list type)

then we are at some form of [System Fω](https://en.wikipedia.org/wiki/Lambda_cube#(F%CF%89)_System_F%CF%89).

We can then restrict it a bit, since we don't need type qualification anywhere. We would only need to allow type qualification at the outermost level, both for types and for terms.

This would give use some form of the  Hindley Milner type system. I was trying to do some research online/in books, and I think this is the same as [prenex polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism#Rank-1_(predicative)_polymorphism)?

## Governance

Chatted with @mwillsey about governance structure of egg and egglog. We talked a bit about the goals of wanting to have a project that students could easily use for their research and build independent work on top of. I mentioned a bit about my recent experience learning about and using the sociocratic form of governance and followed up by [posting links](https://egraphs.zulipchat.com/#narrow/stream/328972-general/topic/governance/near/368047631) to an [intro talk](https://www.youtube.com/watch?v=bYfm3Nn6IbY) and [paper](https://scholarsarchive.byu.edu/cgi/viewcontent.cgi?article=1038&context=joni) on the subject.
  

# Other Talks/chats
Found this link in my notes but forget what its for! [http://adam.chlipala.net/](http://adam.chlipala.net/)

  
  

## Parallel Patterns

- [Calyx](https://calyxir.org/) - IR for hardware design

- Had [Tutorial](https://calyxir.org/tutorial/) on how to use it PLDI which can follow along with

- [Pollen](https://github.com/cucapra/pollen) - Compiles to Calyx, higher level language for genomics

- [Sparse Abstract Machine](https://arxiv.org/abs/2208.14610) - Like TACO, but compiles to parallel hardware

  

## Different conferences?

I mentioned that lots of the PLDI talks seemed to include performance numbers and have a lot of focus on implementation. Someone mentioned that [ICFP (International Conference on Functional Programming)](https://www.icfpconference.org/) has more theory and that [POPL (Principles of Programming Languages)](https://www.sigplan.org/Conferences/POPL/) is "the oscars" that you might send your better papers there (not trying to start a flame war or anything, just my first time at the conference and trying to understand the landscape).

  

## [Indexed Streams](https://cutfree.net/PLDI_2023_indexed_streams.pdf)

Talk on formal semantics for iterative data structures. Similar in some ways again to TACO semantics. @kovach has an implementation in lean called [etch](https://github.com/kovach/etch/).

I loved the talk, great slides that were all drawn with a tablet. I didn't get a full grasp of all the math. From my perspective, seemed similar to an abstract semantics of like an iterable in Python, but with an added skip operator to pass as well as another operator to allowed some linear time indexing in the next N elements. This allows it to represent things which support fast forward indexing like vectors, but also things like linked lists or dynamic streams.

I would like to read the paper and see if I can get a better handle on it by doing out some simple examples by hand. It seems like the type old structure would could make a nice IR in egg for holding ndarrays, capturing some subset of dense and sparse arrays.

  

## Mosaic Compiler

@manya-bansal gave a great talk on the [Mosaic Compiler](https://manya-bansal.github.io/papers/pldi23main-p107-final.pdf), which can stitch together multiple different sub-compilers and pre-compiled functions to generate code for some tensor expressions. https://github.com/manya-bansal/mosaic

  
<img width="744" alt="Screenshot 2023-06-29 at 12 32 54 PM" src="https://github.com/saulshanabrook/saulshanabrook/assets/1186124/66e98392-ec58-468f-9f1d-6134afd2de84">

It builds on the TACO abstractions can can compile to [Stardust](https://arxiv.org/abs/2211.03251), another IR for hardware.

### Mixed Embeddings

I bit ago I started looking looking using at [interaction trees](https://www.semanticscholar.org/paper/Formal-reasoning-about-layered-monadic-interpreters-Yoon-Zakowski/0f4f20bead15145078286e4334c0f39f7fec6862) (used in Vellvm to try to formally verify LLVM compiler pass) to see if we (with @yifanwu at Linea) could possibly use the same abstraction in an analyzer for Python to reason about its semantics.

From my high level understanding, they provide a way to decompose the different impure pieces of a program, so that you can reason about them separately.

<img width="765" alt="Screenshot 2023-06-29 at 12 20 34 PM" src="https://github.com/saulshanabrook/saulshanabrook/assets/1186124/778d5263-c867-4d1e-9f18-cebbe10c70fb">

I am curious if folks are trying to use the same abstraction, but instead of just using the abstract for to reason about correctness, to then use that IR as way to perform some optimizations and rewrites on, before then lowering back to an imperative form for execution. For example, my intuition is that by separating out things like the memory model from IO side effects, this can lead to some "automatic" optimizations. Of course, compilers already do these passes, but the question is more if have a more formal abstraction of the program as the core, then these passes can be more principles and mathematical.

I am sure this is not a new idea at all, but am looking forward to trying it out more concretely with some of these new abstractions.

I chatted with @lastland about their [great talk on Tlön Embeddings](https://pldi23.sigplan.org/details/pldi-2023-pldi/88/Program-Adverbs-and-Tl-n-Embeddings), which from a high level I understand as related to interaction trees but better support parallel computation.

  

### Tim

I was bothering @alberdingk-thijm about some of these things and they recommended a few other folks to look into:

* [Frank Pfenning at CMU](http://www.cs.cmu.edu/~fp/)

* [Matthew Weaver at Princeton](https://www.cs.princeton.edu/~mzweaver/)

* His advisor [Dan Licata at Wesleyan](https://dlicata.wescreates.wesleyan.edu/)

  
## Talia Ringer
I was talking to @ezrosent about [Vladimir Voevodsky](https://www.math.ias.edu/vladimir/home) and the relationship between type theory, mathematics, and computation, and was asking for resources on some of the above questions about if recent advances in theorem proving/dependent type software could be used in compilers, not with the goal of proving their correctness per-se, but more so to use the same abstractions for optimization at a higher level of semantics.

He suggested I bug mastadon about it (which [I tried](https://social.coop/@saul/110582254469319027)) and also maybe I could talk @tlringer who was there at the conference.

They were very generous with their time and gave me a bunch of possible resources over lunch.

Here are some of the notes:
* [Alvin Cheung](https://people.eecs.berkeley.edu/~akcheung/)'s work on verified lifiting
* Check out who cites the [Cogent](https://www.semanticscholar.org/paper/Cogent%3A-Verifying-High-Assurance-File-System-Amani-Hixon/948837266d3df839d4e3e1bc4d63b90e2fc9cae9) verified compiler project
* [Martin Rinard](https://people.csail.mit.edu/rinard/)
* Their own lecture notes and resources on [e-graphs with dependent types](https://dependenttyp.es/classes/readings/17-egraphs.html) from [their proof automation class](https://dependenttyp.es/classes/598sp2022.html)
* They also mentioned the work in homotopy type theory. In particular they found the concept of path equality in cubical type theory to be very beautiful and interesting. For learning more about that they recommended a discord (Maybe the one on [1lab](https://1lab.dev/)?)

**Comment by tonyfast** [comment]
you think the wasm implementation will be hard?

**Reply by saulshanabrook** [reply]
It seems doable, but I haven't tried it?

This is the path I started to chart to get interactive notebooks working:

1. I use the [MySt NB extension](https://myst-nb.readthedocs.io/en/latest/) for Sphinx to render notebooks.
2. Using [sphinx-thebe](https://sphinx-thebe.readthedocs.io/en/latest/examples/notebooks.html) we can add interactive notebooks
3. Thebe [seems to have added supported for JupyterLite](https://mystmd.org/thebe/quickstart-lite), which can load other wheels
4. Pyodide has [some instructions on how to build custom wheels ](https://pyodide.org/en/stable/development/building-and-testing-packages.html#building-binary-packages-for-pyodide). There is [a blog post from a year ago](https://blog.pyodide.org/posts/rust-pyo3-support-in-pyodide/) about getting it working with rust!

If you want to try messing around with it, happy to help.
